### PR TITLE
Add `/api/ctf/start` and `/api/ctf/stop` routes

### PR DESCRIPTION
**Pull request by mradigen**
This PR adds the following:
- Starting and stopping of Docker containers
- Allow for response messages to be configured from `config.py`
- Tortoise Models for Container and CTF
- Restructure DB to run asynchronously with FastAPI without multiple event loops
- Update docs

I could not add tests cause I was unable to figure out how to run asynchronous tests with pytest.

### CODE REVIEW

**Comment by mradigen** [comment]
Not sure why mypy is considering `image_config["ports"]` as a `str` when it is a `dict`. Additionally it seems that the docker module doesn't have a `py.typed`, hence mypy throws an error for that.

File: README.md [hunk 1]
```diff
@@ -12,7 +12,8 @@
 ```sh
 python -m venv .venv                # Create a python virtual environment
 source .venv/bin/activate           # Activate it (This command will differ for Windows)
-pip instal -r requirements.txt      # Install the dependencies
+pip install poetry
```
Comment (WizzyGeek) [inline README.md:15]: poetry **must** not be installed in the venv, but globally, see https://python-poetry.org/docs/#installation

File: src/pwncore/routes/ctf/start.py [hunk 2]
```diff
@@ -1,20 +1,168 @@
...
+from fastapi import Response
+import uuid
+from tortoise.transactions import atomic
+
 from pwncore.routes.ctf import router
+from pwncore.db import Container, CTF
+from pwncore.container import docker_client
+from pwncore.config import DEV_CONFIG
+
+config = DEV_CONFIG
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/start.py:12]: I didn't intend to export `DEV_CONFIG` can you write this same line in config.py

File: src/pwncore/db.py
Comment (WizzyGeek) [inline src/pwncore/db.py:1]: Layout it out as a subpackage 
`pwncore.models`

```
src/pwncore/models/__init__.py
src/pwncore/models/model1.py
src/pwncore/models/model2.py
```

in `__init__.py` define `__all__` which contains names of all Tortoise ORM and Pydantic models

for any model related methods, just define it in the model as a method, otherwise lay it in `pwncore.utils`

File: src/pwncore/routes/ctf/start.py [hunk 3]
```diff
@@ -1,20 +1,168 @@
...
+async def stop_docker_container(ctf_id: int, response: Response):
+
+    ctf = await CTF.get_or_none(id=ctf_id)
+    if not ctf:
+        response.status_code = 404
+        return {"msg": config.messages["ctf_not_found"]}
+
+    team_id = get_team_id()
+    team_container = await Container.get_or_none(team_id=team_id, ctf_id=ctf_id)
+    print(team_container)
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/start.py:150]: Same as before, use debug

File: src/pwncore/routes/ctf/start.py [hunk 4]
```diff
@@ -1,20 +1,168 @@
...
+    # try:
+    await Container.filter(team_id=team_id, ctf_id=ctf_id).delete()
+    # except:  # Not sure which exception should be filtered here for
+    #     response.status_code = 500
+    #     return {
+    #         "msg": config.messages["db_error"]
+    #     }
 
+    container = docker_client.containers.get(team_container.id)
+    container.stop()
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/start.py:165]: Are these calls blocking?

**Review by WizzyGeek — CHANGES_REQUESTED** [review]

File: src/pwncore/db.py
Comment (mradigen) [reply src/pwncore/db.py:1]: That seems to over complicate our structure, all our models combined would be < 150 LoC.

File: src/pwncore/routes/ctf/start.py (hunk 4 above)
Comment (mradigen) [reply src/pwncore/routes/ctf/start.py:165]: Yes they are, we could add a timeout to stop if it does not stop in the mentioned time (very unlikely since we're running basic programs on the docker container)

File: src/pwncore/routes/ctf/start.py (hunk 3 above)
Comment (mradigen) [reply src/pwncore/routes/ctf/start.py:150]: My bad, that was just a line for debugging the DB. I'll remove it

File: src/pwncore/routes/ctf/start.py (hunk 4 above)
Comment (WizzyGeek) [reply src/pwncore/routes/ctf/start.py:163]: Please move all blocking **IO** calls in the default Thread pool executor

more info @ https://docs.python.org/3/library/asyncio-eventloop.html?highlight=run_in_executor#asyncio.loop.run_in_executor

The reason is that blocking io causes the eventloop to lockup, hence hurts concurrency

TLDR, breaks the scheduling algorithm, move it in a thread pool

File: src/pwncore/db.py
Comment (WizzyGeek) [reply src/pwncore/db.py:1]: There's two reasons I want this.

1. Our database is denormalised right now, as I move to just 2NF
Multiple tables will be created.

2. The current models.py in the old project is difficult to navigate atleast for me
I often work with multiple editors, and this layout plays good with that.

I judge this to be a trade worth the marginal code duplication and complexity gained

File: src/pwncore/routes/ctf/__init__.py [hunk 5]
```diff
@@ -11,6 +11,7 @@
 
 router = APIRouter(prefix="/ctf", tags=["ctf"])
 
+from . import start
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/__init__.py:14]: Why are you using dependency injection internally?
Is there no better way to do this?

Comment (WizzyGeek) [reply src/pwncore/routes/ctf/__init__.py:14]: You can invert this relation with a non-prefixed router
https://fastapi.tiangolo.com/reference/apirouter/?h=api#fastapi.APIRouter--example

File: src/pwncore/routes/ctf/start.py (hunk 4 above)
Comment (WizzyGeek) [reply src/pwncore/routes/ctf/start.py:163]: Ideal solution is this https://github.com/aio-libs/aiodocker
or some other API wrapper which is already async

File: src/pwncore/db.py
Comment (mradigen) [reply src/pwncore/db.py:1]: Got it. I'll anyways be removing the db.py altogether once #7 is merged.

File: pyproject.toml [hunk 6]
```diff
@@ -19,6 +19,8 @@ from = "src"
 python = "^3.11"
 fastapi = "^0.104.1"
 uvicorn = "^0.24.0"
+tortoise-orm = "^0.20.0"
+aiodocker = "^0.21.0"
```
Comment (WizzyGeek) [inline pyproject.toml:23]: Seems like aiodocker doesn't pass 3.12

but you can try locking aiohttp to https://pypi.org/project/aiohttp/3.9.0rc0/
to pass 3.12, either way the stable wheels for aiohttp on 3.12 should be released soon so we can ignore 3.12 for now

File: src/pwncore/container.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+import atexit
+import asyncio
+
+docker_client = aiodocker.Docker()
+
+
+async def docker_cleanup():
+    await docker_client.close()
+
+atexit.register(lambda: asyncio.run(docker_cleanup()))
```
Comment (WizzyGeek) [inline src/pwncore/container.py:11]: We should register this on the app shutdown

File: src/pwncore/routes/ctf/__init__.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 from fastapi import APIRouter
+from pwncore.routes.ctf import start
```
Comment (WizzyGeek) [inline src/pwncore/routes/ctf/__init__.py:4]: `from pwncore.routes.ctf.start import router as start_router`

because currently it looks like a circular import (less readable)

**Comment by WizzyGeek** [comment]
Also run, `tox run -e black`
In order to fix the type errors you will have to jump through some hoops, it should become easier to pass through after #7

File: pyproject.toml (hunk 6 above)
Comment (mradigen) [reply pyproject.toml:23]: Will wait for the stable wheels for aiohttp then. Leaving as is for now.

**Review by WizzyGeek — APPROVED** [review]
Everything looks good, except the docker client being initialised before the app runs, in the future it might break if someone decides to change the eventloop, but we can ignore that if it works rn

**Review by KreativeThinker — APPROVED** [review]

**Review by nexxeln — APPROVED** [review]
